File: tools/progression-sheets.mjs
#!/usr/bin/env node
// tools/progression-sheets.mjs — generates the Shipments/Orders search-progression
// sibling deliverables (xlsx + csv) FROM THE CODE, so they cannot drift from what
// is deployed. See docs/superpowers/plans/progression-sheets-sibling-deliverable.md.
//
// MUST be run from the repo ROOT (that is where `vite` is hoisted). Loads the
// search modules through Vite's own resolver (ssrLoadModule) because
// src/search/** uses extensionless imports and pulls in a .ts file that plain
// `node` cannot import — do not substitute a node import here.
//
// Single-sheet "Progression Grouping" design (approved v4, user, 2026-09-11).
// Both the vault master and the docs/story-packs copy are the SAME workbook —
// written once by the Python renderer, then copied — so they cannot diverge.
// The vault CSV is written by the same Python renderer, from the same computed
// rows as the xlsx, so csv and workbook cannot disagree either.
//
// Usage:
//   node tools/progression-sheets.mjs           write the 6 committed outputs
//   node tools/progression-sheets.mjs --check   regenerate into a temp dir and
//                                                byte-compare against committed
import { createServer } from 'vite'
import { readFileSync, writeFileSync, mkdtempSync, rmSync, mkdirSync, existsSync, copyFileSync } from 'node:fs'
import { spawnSync } from 'node:child_process'
import { tmpdir } from 'node:os'
import path from 'node:path'

const ROOT = process.cwd()
const APP_ROOT = 'apps/odyssey-one'
const PY = '/tmp/pptx_env/bin/python'
const RENDERER = path.join(ROOT, 'tools/progression-sheets.py')
const PROSE_PATH = path.join(ROOT, 'docs/progression/prose.json')
const CHECK = process.argv.includes('--check')

const OUT = {
  shipmentsVaultXlsx: 'vault/10-domains/shipments/data/attributes-progression-grouping.xlsx',
  ordersVaultXlsx: 'vault/10-domains/orders/data/attributes-progression-grouping.xlsx',
  shipmentsStoryXlsx: 'docs/story-packs/shipments-search-progression.xlsx',
  ordersStoryXlsx: 'docs/story-packs/orders-search-progression.xlsx',
  shipmentsCsv: 'vault/10-domains/shipments/data/attributes-progression-grouping.csv',
  ordersCsv: 'vault/10-domains/orders/data/attributes-progression-grouping.csv',
}

// ── Load code truth via Vite (verified working — do not substitute) ────────
async function loadCodeTruth() {
  const server = await createServer({
    root: APP_ROOT,
    configFile: `${APP_ROOT}/vite.config.js`,
    server: { middlewareMode: true },
    logLevel: 'error',
  })
  try {
    const shipProg = await server.ssrLoadModule('/src/search/shipments/progression.js')
    const shipCriteria = await server.ssrLoadModule('/src/search/shipments/criteria.js')
    const ordProg = await server.ssrLoadModule('/src/search/orders/progression.js')
    const ordReg = await server.ssrLoadModule('/src/search/orders/registry.js')
    return { shipProg, shipCriteria, ordProg, ordReg }
  } finally {
    await server.close()
  }
}

// ── Panel-control derivation ────────────────────────────────────────────────
// Shipments: derived from attr.match, exactly as ShipmentsFiltersView.renderControl does.
function shipmentsPanelControl(match) {
  if (match === 'enum') return 'enum chips (multi-select)'
  if (match === 'letters') return 'combobox (typable, lazy in live)'
  if (match === 'date') return 'date picker + date range'
  return 'text field' // digits, both
}

// Orders registry keys that name a different concept than the progression
// dataKey of the same field (only the location pair differs; everything else
// in ORDERS_FILTER_ATTRS shares its `key` with the progression `dataKey`).
const REGISTRY_KEY_TO_PROGRESSION_DATAKEY = {
  origin: 'shipperLocation',
  destination: 'destinationLocation',
}

function buildOrdersPanelMap(ORDERS_FILTER_ATTRS) {
  const map = new Map() // progression dataKey -> registry attr
  for (const reg of ORDERS_FILTER_ATTRS) {
    const dataKey = REGISTRY_KEY_TO_PROGRESSION_DATAKEY[reg.key] ?? reg.key
    map.set(dataKey, reg)
  }
  return map
}

// ── Sidecar validation ──────────────────────────────────────────────────────
function validateSidecar(domain, attrs, prose) {
  const errors = []
  const proseKeys = new Set(Object.keys(prose?.attributes ?? {}))
  const codeKeys = new Set(attrs.map((a) => a.dataKey))
  for (const a of attrs) {
    if (!proseKeys.has(a.dataKey)) {
      errors.push(`${domain}:${a.dataKey} — attribute exists in code but has no prose entry`)
    }
  }
  for (const key of proseKeys) {
    if (!codeKeys.has(key)) {
      errors.push(`${domain}:${key} — prose entry names an attribute key that no longer exists in code`)
    }
  }
  return errors
}

// ── Domain payload builder ──────────────────────────────────────────────────
// Produces exactly the columns the Python renderer's `ix` lookup needs, in the
// same shape as the sheet the design was ported from: {headers, rows}.
const ATTR_HEADERS = ['#', 'Group', 'Suggestions panel header', 'Attribute (bar label)', 'dataKey', 'Match',
  'Exact?', 'Enum values', 'Example (seeded)', 'Panel control', 'Panel label', 'Panel section', 'Free-text',
  'Status', 'Description', 'Notes']

function buildAttrsSheet({ PROGRESSION, ATTRIBUTES, freeTextKeys, panelControlFor, panelSectionFor, prose, notBuiltRows }) {
  const rows = ATTRIBUTES.map((a, i) => {
    const p = prose.attributes[a.dataKey] ?? { description: '', notes: '', example: '' }
    const { control, label } = panelControlFor(a)
    let notes = p.notes || ''
    if (a.match === 'date') {
      const rangeNote = `Also renders a second field labelled "${a.label} Range".`
      notes = notes ? `${notes} ${rangeNote}` : rangeNote
    }
    const suggestionsHeader = PROGRESSION.find((g) => g.group === a.group)?.label ?? ''
    const freeText = freeTextKeys === 'n/a' ? 'n/a' : (freeTextKeys.includes(a.dataKey) ? 'Yes' : '')
    return [i + 1, a.group, suggestionsHeader, a.label, a.dataKey, a.match, a.exact ? 'Yes' : '',
      a.values ? a.values.join(' · ') : '', p.example || '', control, label, panelSectionFor(a) || '',
      freeText, 'Implemented', p.description || '', notes]
  })
  for (const nb of notBuiltRows) {
    rows.push(['', nb.proposedGroup, '', nb.attribute, '', '', '', '', '', '', '', '', '', nb.status, '', nb.why])
  }
  return { headers: ATTR_HEADERS, rows }
}

async function main() {
  const { shipProg, shipCriteria, ordProg, ordReg } = await loadCodeTruth()
  const prose = JSON.parse(readFileSync(PROSE_PATH, 'utf8'))

  const sidecarErrors = [
    ...validateSidecar('shipments', shipProg.SHIPMENTS_ATTRIBUTES, prose.shipments),
    ...validateSidecar('orders', ordProg.ORDERS_ATTRIBUTES, prose.orders),
  ]
  if (sidecarErrors.length) {
    console.error('Sidecar validation failed:')
    for (const e of sidecarErrors) console.error(`  ${e}`)
    process.exit(1)
  }

  // Shipments: panel IS the progression — every attribute is a panel field.
  const shipmentsAttrs = buildAttrsSheet({
    PROGRESSION: shipProg.SHIPMENTS_PROGRESSION,
    ATTRIBUTES: shipProg.SHIPMENTS_ATTRIBUTES,
    freeTextKeys: shipCriteria.FREE_TEXT_KEYS,
    panelControlFor: (a) => ({ control: shipmentsPanelControl(a.match), label: a.label }),
    panelSectionFor: (a) => a.group, // panel IS the progression — same section header (group.group)
    prose: prose.shipments,
    notBuiltRows: prose.shipments.notBuilt,
  })

  const ordersPanelMap = buildOrdersPanelMap(ordReg.ORDERS_FILTER_ATTRS)
  const ordersAttrs = buildAttrsSheet({
    PROGRESSION: ordProg.ORDERS_PROGRESSION,
    ATTRIBUTES: ordProg.ORDERS_ATTRIBUTES,
    freeTextKeys: 'n/a',
    panelControlFor: (a) => {
      const reg = ordersPanelMap.get(a.dataKey)
      return reg ? { control: reg.control, label: reg.label } : { control: '— none', label: '' }
    },
    panelSectionFor: (a) => ordersPanelMap.get(a.dataKey)?.group ?? '',
    prose: prose.orders,
    notBuiltRows: prose.orders.notBuilt,
  })

  const writeDir = CHECK ? mkdtempSync(path.join(tmpdir(), 'progression-sheets-')) : ROOT
  const outPaths = Object.fromEntries(Object.entries(OUT).map(([k, rel]) => [k, path.join(writeDir, rel)]))
  for (const p of Object.values(outPaths)) mkdirSync(path.dirname(p), { recursive: true })

  const pyPayload = {
    shipments: { attrs: shipmentsAttrs, xlsxPath: outPaths.shipmentsVaultXlsx, csvPath: outPaths.shipmentsCsv },
    orders: { attrs: ordersAttrs, xlsxPath: outPaths.ordersVaultXlsx, csvPath: outPaths.ordersCsv },
  }
  const result = spawnSync(PY, [RENDERER], { input: JSON.stringify(pyPayload), encoding: 'utf8' })
  if (result.status !== 0) {
    console.error(result.stderr || result.stdout)
    process.exit(1)
  }
  if (result.stdout) process.stdout.write(result.stdout)

  // The story-pack copy is the SAME workbook as the vault master — copy rather
  // than re-render, so the two files cannot diverge.
  copyFileSync(outPaths.shipmentsVaultXlsx, outPaths.shipmentsStoryXlsx)
  copyFileSync(outPaths.ordersVaultXlsx, outPaths.ordersStoryXlsx)

  if (!CHECK) {
    console.log('Wrote:')
    for (const p of Object.values(OUT)) console.log(`  ${p}`)
    return
  }

  // --check: byte-compare temp dir against committed files.
  let drift = false
  for (const [name, relPath] of Object.entries(OUT)) {
    const committed = path.join(ROOT, relPath)
    const generated = path.join(writeDir, relPath)
    let same = false
    if (existsSync(committed) && existsSync(generated)) {
      same = Buffer.compare(readFileSync(committed), readFileSync(generated)) === 0
    }
    console.log(`${same ? 'ok' : 'DRIFT'}  ${relPath}`)
    if (!same) drift = true
  }
  rmSync(writeDir, { recursive: true, force: true })
  if (drift) process.exit(1)
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: tools/progression-sheets.py
#!/usr/bin/env python3
"""tools/progression-sheets.py — openpyxl renderer for the Shipments/Orders
search-progression workbooks. Reads a JSON payload on stdin (written by
tools/progression-sheets.mjs) and writes, per domain, ONE xlsx (single sheet
"Progression Grouping") and its byte-for-byte-consistent sibling CSV.

This is the approved v4 design (user, 2026-09-11) — see the working prototype
this was ported from. It supersedes the old 7-sheet 'legacy'/'designed'
renderer; there is only one workbook shape now.
"""
import sys
import csv
import json
import re
import io
import datetime
import zipfile
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment
from openpyxl.utils import get_column_letter

ROOT = __import__('pathlib').Path(__file__).resolve().parent.parent
FALLBACK_PATH = ROOT / 'docs/progression/csv-fallback.json'

# Fixed (not "now") so two saves of byte-identical content produce byte-identical
# files — required for `progression:audit` (--check) to byte-compare.
FIXED_DT = datetime.datetime(2026, 1, 1)
ZIP_EPOCH = (1980, 1, 1, 0, 0, 0)


def normalize_zip_timestamps(path):
    # openpyxl's writer forcibly stamps workbook.properties.modified = now() at
    # save() time — core.xml's <dcterms:modified> is patched here too, alongside
    # per-member zip timestamps, so two saves of identical content agree byte-for-byte.
    with open(path, 'rb') as f:
        src = zipfile.ZipFile(f)
        infos = src.infolist()
        contents = {i.filename: src.read(i.filename) for i in infos}
    core = contents.get('docProps/core.xml')
    if core is not None:
        fixed = FIXED_DT.strftime('%Y-%m-%dT%H:%M:%SZ').encode()
        contents['docProps/core.xml'] = re.sub(
            rb'(<dcterms:modified[^>]*>)[^<]*(</dcterms:modified>)',
            lambda m: m.group(1) + fixed + m.group(2),
            core,
        )
    with zipfile.ZipFile(path, 'w', zipfile.ZIP_DEFLATED) as out:
        for i in infos:
            i.date_time = ZIP_EPOCH
            out.writestr(i, contents[i.filename])


def argb(h):
    """6-char hex -> opaque ARGB. openpyxl pads a bare 6-char hex to '00'+hex,
    whose 00 alpha makes Excel render NO fill at all (Quick Look ignores alpha,
    which is why previews lied). Every fill in this file MUST go through this."""
    h = h.lstrip('#')
    return h if len(h) == 8 else 'FF' + h


def solid(h):
    return PatternFill('solid', fgColor=argb(h))


COLS = ['Progression\n(suggestions panel header)', 'Stakeholder Group',
        'Suggested Group\n(search bar grouping)',
        'Attribute', 'Example', 'Description', 'Field Type\n(as built on Vercel)',
        'Filters panel\nfield label', 'Filters panel\nsection', 'Search status', 'Notes']
WID = [26, 24, 24, 24, 26, 38, 22, 26, 22, 19, 42]
NGROUP = 3

RAMP = ['EAF4FC', 'D6EAF8', 'AED6F1', '5DADE2', 'EAF7EA', 'D4EFDF', 'BFE5CC', 'A9DFBF', 'FADBD8', 'F5CBA7']
GREY = 'EDEDED'

# AS BUILT — every value below is the control the deployed app actually renders.
# Keys are the exact `Panel control` strings the generator derives from the code.
AS_BUILT = {
    'text':                             'Text Input',
    'text field':                       'Text Input',
    'combobox':                         'Autocomplete Text (lazy)',
    'combobox (lazy)':                  'Autocomplete Text (lazy)',
    'combobox (typable, lazy in live)': 'Autocomplete Text',
    'location':                         'Autocomplete Text (City/State/Country)',
    'location (lazy)':                  'Autocomplete Text (City/State/Country)',
    'date-range':                       'Date Picker + Range',
    'date picker + date range':         'Date Picker + Range',
    'enum':                             'Dropdown (multi-select)',
    'enum chips':                       'Dropdown (multi-select)',
    'enum chips (multi-select)':        'Dropdown (multi-select)',
    'comparator':                       'Operator + Number',
}
# No panel field at all: the attribute is reachable only by typing in the bar.
BAR_ONLY = {'enum': 'Search bar only — enum value', 'digits': 'Search bar only — number',
            'both': 'Search bar only — text/number', 'letters': 'Search bar only — text',
            'date': 'Search bar only — date'}

NOT_FILTER = 'Not a filter — search bar only'
NO_SECTION = '—'

SEARCH_STATUS = {'Implemented': 'Live in search',
                  'Proposed — not built': 'Not in search',
                  'Skipped': 'Not in search (skipped in proposal)'}

# Verified present on a seeded shipments.json row, but with no search attribute.
DATA_EXISTS = {'Shipment Sequence Leg': 'shipmentSequenceLeg',
               'Next Shipment ID': 'nextShipmentId',
               'Validation Message': 'validationMessage'}


def field_type(r, ix, fallback, a):
    """What Vercel renders for this attribute today. Falls back to the 2026-05
    proposed type ONLY for rows that were never built, where a proposal is all
    that exists and the Status column already says so."""
    if r[ix['Status']] != 'Implemented':
        return fallback['fld'].get(a) or 'Proposed — no control built'
    ctrl = (r[ix['Panel control']] or '').strip().lower()
    if not ctrl or ctrl.startswith('—'):
        return BAR_ONLY.get(r[ix['Match']], 'Search bar only')
    if ctrl not in AS_BUILT:
        raise SystemExit(f'unmapped panel control {ctrl!r} for {a!r} — refusing to guess')
    return AS_BUILT[ctrl]


def short(t, n=110):
    t = re.sub(r'\s+', ' ', (t or '').strip())
    if not t:
        return ''
    if len(t) <= n:
        return t
    cut = t[:n]
    sp = cut.rfind('. ')
    return (cut[:sp + 1] if sp > 40 else cut.rstrip() + '…')


def panel_section(r, ix):
    """Which section of the FILTERS PANEL holds this attribute. Per attribute,
    not per group."""
    return r[ix['Panel section']] or NO_SECTION


def search_status(raw):
    return SEARCH_STATUS.get(raw, raw)


def note_for(a, raw, note):
    if raw != 'Implemented' and a in DATA_EXISTS:
        extra = f"Field `{DATA_EXISTS[a]}` EXISTS on the shipment row; never exposed to search."
        return (extra + ' ' + (note or '')).strip()
    return note


def panel_field(r, ix):
    """What the FILTERS PANEL calls this attribute, or that it has no filter at
    all."""
    ctrl = (r[ix['Panel control']] or '').strip()
    if not ctrl or ctrl.startswith('—'):
        return NOT_FILTER
    return r[ix['Panel label']] or r[ix['Attribute (bar label)']]


def compute_rows(dom_key, attrs, fallback):
    """Groups attrs['rows'] by Group (in first-seen order) and returns, per
    group: (drill_label, stakeholder, group_name, [computed attribute rows]).
    Each computed attribute row is the 11 values for columns D..K (Attribute..
    Notes) — the shared truth both the xlsx and the csv render from."""
    H = attrs['headers']
    ix = {k: H.index(k) for k in ['Group', 'Suggestions panel header', 'Attribute (bar label)',
          'Example (seeded)', 'Description', 'Match', 'Panel control', 'Status', 'Notes',
          'Panel section', 'Panel label']}
    order, blocks = [], {}
    for r in attrs['rows']:
        g = r[ix['Group']]
        if g not in order:
            order.append(g)
            blocks[g] = []
        blocks[g].append(r)

    groups = []
    for gi, g in enumerate(order):
        rows = blocks[g]
        built = any(r[ix['Status']] == 'Implemented' for r in rows)
        attr_rows = []
        for r in rows:
            a = r[ix['Attribute (bar label)']]
            ft = field_type(r, ix, fallback, a)
            desc = r[ix['Description']] or fallback['desc'].get(a, '')
            attr_rows.append([
                a, r[ix['Example (seeded)']], short(desc), ft,
                panel_field(r, ix), panel_section(r, ix),
                search_status(r[ix['Status']]),
                short(note_for(a, r[ix['Status']], r[ix['Notes']]), 110),
            ])
        drill = rows[0][ix['Suggestions panel header']] or ''
        if dom_key == 'shipments':
            stk = fallback['stk'].get(rows[0][ix['Attribute (bar label)']], 'n/a — grid-column derived')
        else:
            stk = 'n/a — grid-column derived'
        if gi == 0:
            drill += '\n(What is it?)'
        arrow = '' if gi == len(order) - 1 else '\n↓'
        groups.append({'drill': drill + arrow, 'stk': stk, 'group': g, 'built': built, 'rows': attr_rows})
    return groups


def write_xlsx(groups, path):
    wb = openpyxl.Workbook()
    ws = wb.active
    ws.title = 'Progression Grouping'
    wb.properties.created = FIXED_DT
    wb.properties.modified = FIXED_DT

    for j, h in enumerate(COLS):
        c = ws.cell(1, j + 1, h)
        c.fill = solid('D9D9D9')
        c.font = Font(name='Calibri', bold=True, size=11)
        c.alignment = Alignment(horizontal='center', vertical='center')
    ws.row_dimensions[1].height = 22

    row = 2
    for gi, grp in enumerate(groups):
        fill = solid(RAMP[gi % len(RAMP)] if grp['built'] else GREY)
        start = row
        for vals in grp['rows']:
            for j, v in enumerate(vals):
                col = j + NGROUP + 1
                c = ws.cell(row, col, v)
                c.fill = fill
                c.font = Font(name='Calibri', size=11)
                c.alignment = Alignment(vertical='center', wrap_text=col in (6, 11),
                                         horizontal='center' if col == 10 else 'left')
                if col in (8, 9) and v in (NOT_FILTER, NO_SECTION):
                    c.font = Font(name='Calibri', size=11, italic=True, color=argb('808080'))
            row += 1
        end = row - 1
        for j, text in ((1, grp['drill']), (2, grp['stk']), (3, grp['group'])):
            if end > start:
                ws.merge_cells(start_row=start, start_column=j, end_row=end, end_column=j)
            c = ws.cell(start, j, text)
            c.fill = fill
            c.font = Font(name='Calibri', size=11, bold=j in (2, 3))
            c.alignment = Alignment(horizontal='center', vertical='center', wrap_text=True)

    for j, w in enumerate(WID):
        ws.column_dimensions[get_column_letter(j + 1)].width = w
    ws.freeze_panes = 'E2'
    ws.page_setup.orientation = 'landscape'
    ws.sheet_properties.pageSetUpPr.fitToPage = True
    ws.page_setup.fitToWidth, ws.page_setup.fitToHeight = 1, 0
    ws.print_title_rows = '1:1'
    wb.save(path)
    normalize_zip_timestamps(path)
    return row - 2


def write_csv(groups, path):
    # Mirrors the merge semantics of the xlsx: columns A-C (Progression,
    # Stakeholder Group, Suggested Group) are blank on every row after a
    # group's first, matching the pre-regeneration CSV precedent this sheet
    # descends from.
    buf = io.StringIO()
    w = csv.writer(buf)
    w.writerow(COLS)
    for grp in groups:
        for i, vals in enumerate(grp['rows']):
            lead = [grp['drill'], grp['stk'], grp['group']] if i == 0 else ['', '', '']
            w.writerow(lead + vals)
    with open(path, 'w', newline='', encoding='utf-8') as f:
        f.write(buf.getvalue())


def build(dom_key, attrs, fallback, xlsx_path, csv_path):
    groups = compute_rows(dom_key, attrs, fallback)
    n = write_xlsx(groups, xlsx_path)
    write_csv(groups, csv_path)
    return n


def main():
    payload = json.load(sys.stdin)
    all_fallback = json.load(open(FALLBACK_PATH))
    for dom_key in ('shipments', 'orders'):
        d = payload[dom_key]
        n = build(dom_key, d['attrs'], all_fallback[dom_key], d['xlsxPath'], d['csvPath'])
        print(f"wrote {d['xlsxPath']}  ({n} rows)")
        print(f"wrote {d['csvPath']}")


if __name__ == '__main__':
    main()
